Revise Backends Config Page

laurapacilio opened the pull request:
This PR:
- Combines the backends overview page with the backend configuration page. The Overview page was redundant in several areas with the configuration page and wasn't necessary.
- Adds a section to the configuration page explaining that users should be careful to supply credentials and other sensitive data using environment variables. If these values are hardcoded into the configuration, Terraform will print them in plain text to both the state file and each plan file.
- Adds warnings to all of the relevant pages that folks should be careful with sensitive credentials data and links them to the explanation on the configuration page for details.

Extends the work started by the community in this PR: https://github.com/hashicorp/terraform/pull/29519#pullrequestreview-998851540
** If accepted, recommendation will be to close that PR and go with this one, as it's more comprehensive by adding warnings to all backend pages, not just HTTP.

---------

Separate PR to redirect the deleted overview page: https://github.com/hashicorp/terraform-website/pull/2343

apparentlymart commented on website/docs/language/settings/backends/configuration.mdx:
```diff
@@ -35,6 +43,18 @@ There are some important limitations on backend configuration:
+
+Backends store state in a remote service, which allows multiple people to access it. Accessing remote state generally requires access credentials, since state data contains extremely sensitive information.
+
+!> **Warning:**  We recommend using environment variables to supply credentials and other sensitive data. If you use `-backend-config` or hardcode these values directly in your configuration, Terraform will include these values in both state and plan files. This can leak sensitive credentials.
```
I broadly like this way of presenting it, but it isn't technically correct to refer to `.terraform/terraform.tfstate` as a "state file" in the usual sense we mean when we say that. I don't know that we really have a good name for that file because we haven't really needed to talk about it in the documentation before.

What do you think about this alternative for the last two sentences?

> If you use `-backend-config` or hardcode these values directly in your configuration, Terraform will include these values in both the `.terraform` subdirectory and in plan files. This can leak sensitive credentials.

Here I referred to the `.terraform` directory broadly, since I expect that's something users will either already be familiar with or will become accustomed to quickly when using Terraform, whereas the exact contents of that directory are not so relevant to everyday Terraform use because `terraform init` (along with some other commands) manages it all automatically.

I think it's fine to refer to the specific path `.terraform/terraform.tfstate` below where we have room for more words to say a little about what that file is, but I don't think it's necessary to go into that much detail in the little callout box we're going to copy onto every backend page.

laurapacilio replied:
I like it! Let me push up a change where I fix all the little boxes with this nice new sentence. And yeah that makes sense to just give the general warning without going into very specific details in the callout. Thank you very much for explaining!

apparentlymart commented on website/docs/language/settings/backends/configuration.mdx:
```diff
@@ -35,6 +43,18 @@ There are some important limitations on backend configuration:
+- The `.terraform/terraform.tfstate` file contains the backend configuration for the current working directory.
+- All plan files capture the information in `.terraform/terraform.tfstate` at the time the plan was created. This helps ensure Terraform is applying the plan to correct set of infrastructure.
+
+Configuration stored in a plan file takes precedence over other configuration. This may cause authentication issues if credentials change between creating and applying plans. For example, this can occur in automated environments where credentials are scoped to the duration of a job.
```
I wonder if this paragraph is currently assuming too much about what the reader is aware of at this point in their learning, talking about "a plan file" without any other context.

Perhaps I'm overthinking it, but I might consider the following in the hope of spelling out a bit more clearly what we're talking about:

> When applying a plan that you previously saved to a file, Terraform will use the backend configuration stored in that file instead of the current backend settings. If that configuration contains time-limited credentials then they may expire before you finish applying the plan. Pass credentials using environment variables when you need to use different values between the plan and apply steps.

laurapacilio replied:
you're not overthinking it! This is an excellent point, and I"m adding this to the docs :-) Thank you!

apparentlymart commented on website/docs/language/settings/backends/configuration.mdx:
```diff
@@ -4,19 +4,27 @@ page_title: Backend Configuration - Configuration Language
+
+This page describes how to configure a backend by adding the [`backend` block](#using-a-backend-block) to your configuration.
+
+-> **Note:** In Terraform versions before 1.1.0, HashiCorp classified backends as standard or enhanced. The enhanced label differentiated the [`remote` backend](/language/settings/backends/remote), which could both store state and perform Terraform operations. This classification has been removed. Refer to [Using Terraform Cloud](/cli/cloud) for details about storing state, executing remote opera ...
```
This is perhaps just an irrelevant nit, but I've not seen many examples of us talking about "HashiCorp" in the third person in this sort of documentation before; we've typically said "we" instead. I guess that might feel a bit funny here, but saying tat "HashiCorp" as a whole did anything feels a bit funny too; maybe "The Terraform team" or "this documentation" would be reasonable compromises if "we" isn't right here? :thinking:

laurapacilio replied:
yeah I went back and forth on this.  I wanted to take it out of passive voice, which it was originally (that avoided this problem entirely, and is why I bet it was written that way). I was thinking about how this note is largely for people who are studying for the HashiCorp certification exam so they're not confused, and that's why I went "HashiCorp". It's not technically wrong, but I agree it's not a convention we typically use. 

You know what? I think I'm going to go back to "we". We use "we" throughout and it's also correct (we, being the teams at HashiCorp who know about such things). Thank you for the thoughtful review, Martin!

laurapacilio commented:
okay @apparentlymart thank you for your review! I believe I addressed all of your suggestions, and I think it's ready for official review and approval :)

njdart commented:
This all looks good to me, happy for you to merge 👍 I'll close my MR
